collector/filesystem_common.go

```diff
@@ -126,7 +127,15 @@ func (c *filesystemCollector) Update(ch chan<- prometheus.Metric) (err error) {
 	if err != nil {
 		return err
 	}
+	// Make sure we expose a metric once, even if there are multiple mounts
+	updated := map[string]bool{}
 	for _, s := range stats {
+		k := fmt.Sprintf("%q", s)
```

Comment:
First of all, `%q` is the printf verb for strings, what you want here is `%v` or `%+v`. Go vet should complain about this actually.

Creating a string key for every stats will require additional allocations. Ideally we would just use the struct itself as key. Looking at `filesystemStats` the only field preventing this is `labelValues` as structs can only be used as keys if their fields are basic types. Looking at labelValues it appears that using a string slice is a weak data type anyway, as there is strong assumption about the order, length and value. I'd suggest to make the assumption explicit and replace `labelValues` with three explicit string fields `"device", "mountpoint", "fstype"`.

This will ensure that OS specific implementations can't pass the values in a wrong order and allow you to use the struct as key, avoiding any further allocations by this addition.